Lib/glyphsLib/glyphdata.py

```diff
@@ -31,69 +31,224 @@
-            prodname = name
-            name = rev_prodname
-    if prodname is None:
-        prodname = name
-    unistr = data.IRREGULAR_UNICODE_STRINGS.get(name)
-    if unistr is None:
-        unistr = agl.toUnicode(prodname)
-    if unistr != "" and name not in data.MISSING_UNICODE_STRINGS:
-        unistr_result = unistr
+            production_name = base_name
+            base_name = rev_prodname
+
+    # Finally, if we couldn't find a known production name one way or another,
+    # conclude that the glyph name doesn't carry any Unicode semantics. Use the
+    # bare name in that case.
+    if production_name is None:
+        production_name = glyph_name
+
+    # Next, derive the actual character from the production name, e.g.
+    # "uni0414" -> "Д". Two caveats:
+    # 1. For some glyphs, Glyphs does not have a mapped character even when one 
+    #    could be derived.
+    # 2. For some others, Glyphs has a different idea than the agl module.
+    characters = None
+    if base_name not in data.MISSING_UNICODE_STRINGS:  # 1.
+        characters = data.IRREGULAR_UNICODE_STRINGS.get(base_name)  # 2.
+        if characters is None:
+            characters = agl.toUnicode(production_name) or None
+
+    # Lastly, generate the category in the sense of Glyph's
+    # GSGlyphInfo.category and .subCategory.
+    category, sub_category = _get_category(base_name, characters, data)
+
+    return Glyph(glyph_name, production_name, characters, category, sub_category)
+
+
+def _lookup_production_name(glyph_name, data=glyphdata_generated):
+    """Return the production name for a glyph name from the GlyphsData.xml
+    database according to the AGL specification.
+
+    Handles single glyphs (e.g. "brevecomb") and ligatures (e.g.
+    "brevecomb_acutecomb"). Returns None when a valid and semantically
+    meaningful production name can't be constructed or when the AGL
+    specification would be violated, get_glyph() will use the bare glyph
+    name then.
+
+    Note:
+    - Glyph name is the full name, e.g. "brevecomb_acutecomb.case".
+    - Base name is the base part, e.g. "brevecomb_acutecomb"
+    - Suffix is e.g. "case".
+    """
+
+    def is_unicode_u_value(name):
+        return name.startswith("u") and all(
+            part_char in "0123456789ABCDEF" for part_char in name[1:]
+        )
+
+    base_name, dot, suffix = glyph_name.partition(".")
+
+    # First, look up the full glyph name and base name in the AGLFN and in
+    # PRODUCTION_NAMES.
+    if (
+        glyph_name in agl.AGL2UV
+        or base_name in agl.AGL2UV
+        or glyph_name in (".notdef", ".null")
+    ):
+        return glyph_name
+
+    if glyph_name in data.PRODUCTION_NAMES:  # e.g. ain_alefMaksura-ar.fina -> uniFD13
+        return data.PRODUCTION_NAMES[glyph_name]
+    if base_name in data.PRODUCTION_NAMES:
+        final_production_name = data.PRODUCTION_NAMES[base_name] + dot + suffix
+        if len(final_production_name) > 31:
```

Comment:
are you sure about the length limit of 31? I remember it being 64, but maybe i'm wrong.

Reply:
https://github.com/adobe-type-tools/agl-specification

> Length and character set limitations on names Glyphs from Western OpenType/CFF and TrueType fonts must still be referenced in many workflows as name-keyed font data. As a result, glyph names are still subject to the length and character set limitations that are imposed by the Type 1 specification and PostScript interpreter implementations. Both of these require that a glyph name be no longer than 31 characters in length [...]

Reply:
ok.. but the Adobe Feature File spec says 63...
https://github.com/adobe-type-tools/afdko/blob/develop/docs/OpenTypeFeatureFileSpecification.html

Reply:
I guess the 64th is a C-string NULL terminating byte :smile:

Reply:
I mean, if makeotf allows 64, who are we to restrict it to 31?!

Reply:
Hmm... so, which one should it be? 😕 

@miguelsousa ?

Reply:
Ok, let's just assume everything is broken anyway and if makeotf is doing it, so should we 😁👍

Reply:
yes, let's use 63 and paste a link to that Feature File spec paragraph in an inline comment, thanks